flexget/utils/serialization.py

```diff
@@ -0,0 +1,128 @@
+import datetime
+from abc import ABC, abstractmethod
+
+from flexget.utils import json
+
+DATE_FMT = '%Y-%m-%d'
+ISO8601_FMT = '%Y-%m-%dT%H:%M:%SZ'
+
+
+def serialize(value):
+    """
+    Convert an object to JSON serializable format.
+    """
+    if isinstance(value, Serializable):
+        return value.serialize()
+    if isinstance(value, (list, tuple)):
+        return [serialize(v) for v in value]
+    if isinstance(value, dict):
+        return {k: serialize(v) for k, v in value.items()}
+    if isinstance(value, datetime.datetime):
+        return DateTimeSerializer.serialize(value)
+    if isinstance(value, datetime.date):
+        return DateSerializer.serialize(value)
+    if isinstance(value, (str, int, float, type(None))):
+        return value
+    raise TypeError(f'`{value!r}` of type {type(value)!r} is not serializable')
+
+
+def deserialize(value):
+    if isinstance(value, dict):
+        if all(key in value for key in ('serializer', 'version', 'value')):
+            return _registry()[value['serializer']].deserialize(value)
+        return {k: deserialize(v) for k, v in value.items()}
+    if isinstance(value, list):
+        return [deserialize(v) for v in value]
+    return value
+
+
+class Serializable(ABC):
+    @abstractmethod
+    def _serialize(self):
+        """This method should be implemented to return a plain python datatype which is json serializable."""
+        pass
+
+    def serialize(self):
+        """Returns a json serializable form of this class, with all information needed to deserialize."""
+        return {
+            'serializer': self.serializer_name(),
+            'version': self.serializer_version(),
+            'value': self._serialize(),
+        }
+
+    @classmethod
+    @abstractmethod
+    def _deserialize(cls, data, version):
+        """Returns an instance of this class, recreated from the serialized form."""
+        pass
+
+    @classmethod
+    def deserialize(cls, data):
+        return cls._deserialize(data['value'], data['version'])
+
+    @classmethod
+    def serializer_name(cls):
+        return cls.__name__
+
+    @classmethod
+    def serializer_version(cls):
+        return 1
+
+    def dumps(self):
+        return json.dumps(self.serialize())
+
+    @classmethod
+    def loads(cls, data):
+        return cls.deserialize(json.loads(data))
+
+
+class DateSerializer(Serializable):
+    @classmethod
+    def serializer_name(cls):
+        return 'date'
+
+    def _serialize(self):
+        pass
+
+    @classmethod
+    def serialize(cls, value):
```

Comment:
Method signature does not match base method. Changing method signature in subclass is bad practice. Type hinting/argument naming would make this more readable as well. I think `value` here refers to a entry field, so why not call the argument `field`. Or even better `datetime: field` if it only serializes that type of fields.

Reply:
Yeah, I went a bit back and forth on how to implement this. It doesn't work quite the same as other things, because it's for serializing a bulitin type, which we don't directly subclass like we do for our own custom types. I sort of emulated how the custom types work so that the deserialization didn't need any custom handling.